### PR TITLE
feat: support `devModules` option

### PR DESCRIPTION
**Pull request by galvez**
As per discussion in #core -- @pi0 does this look right to you?

### CODE REVIEW

File: packages/core/src/module.js [hunk 1]
```diff
@@ -16,6 +16,11 @@ export default class ModuleContainer {
     // Call before hook
     await this.nuxt.callHook('modules:before', this, this.options.modules)
 
+    if (this.options._start) {
```
Comment (pi0) [inline packages/core/src/module.js]: I think this condition should be like `if (!...)`

Comment (pi0) [reply packages/core/src/module.js]: PS: Seems good idea to run devModules **before** modules :+1:

Comment (galvez) [reply packages/core/src/module.js]: Ops, updated the check -- will try and add a test next :)

**Comment by manniL** [comment]
I really like the idea but I'm not sure if it should be implemented that way, especially because if we want to change it later, we'd have to wait for the next major.

Looking at https://github.com/nuxt/rfcs/issues/10, how will this fit in the concept there? 🤔 Shouldn't we discuss this idea in the RFC / as part of the RFC first?

Wouldn't it be nice to give the module itself an option to automatically register only in dev/prod/both?

Not sure if an extra array in the config this the right approach for it. 
Maybe an option instead?

**Comment by pi0** [comment]
@manniL I taught about module improvement RFC as well. But this is a different matter. A module to be able to decide to register or not have to be _installed_ first which means have to be always in "dependencies" section of package.json.

**Comment by manniL** [comment]
@pi0 I know what you mean, yes. I solved this in the past via checking NODE_ENV manually (rel: https://twitter.com/TheAlexLichter/status/1085179308606349314/photo/1).

Anyway, having `modules` and `devModules` feels a bit redundant. Might be just me though 🤷‍♂️

**Comment by pi0** [comment]
@manniL NODE_ENV of `nuxt build` and `nuxt start` is both `production` :)

**Comment by manniL** [comment]
> @manniL NODE_ENV of `nuxt build` and `nuxt start` is both `production` :)


Right, but not for `nuxt dev` (which I thought what that feat it was for).

So, I hope I get this right now:

`devModules` -> modules that are need for `build` (no matter if  `nuxt dev` or `nuxt generate` or `nuxt build`) but not needed for `nuxt start`?

**Comment by pi0** [comment]
Yeah. Basically `modules`/`devModules` is mapped to `dependencies`/`devDependencies`. Modules like `@nuxt/pwa` which are only needed for build/dev are inside `devDependencies` and `devModules` :) BTW I see your point.

**Comment by manniL** [comment]
@pi0 Ahh, got it now! That makes sense but is confusing when it comes to "running in dev/prod mode"

**Comment by pi0** [comment]
I think would be better to add it as a non documented feature for now. Modules that support this mode (pwa for example breaks for spa) can advice users to use correct section. Still I'm open to better wording for devModules :)

**Review by pi0 — DISMISSED** [review]

**Comment by manniL** [comment]
I thought about using a nested object but that doesn't seem sexy as well :D

```js
{
  modules: {
    dev: [],
    prod: [],
    all: [] // or so?
  }
}
```

**Comment by pi0** [comment]
@manniL is really more confusing. Also somehow breaking change without patch on `options.js`

**Comment by manniL** [comment]
@pi0 Yeah, that was my thought too.

**Comment by Atinux** [comment]
Side note: We should also be able to handle it here when importing modules: https://github.com/nuxt/nuxt.js/blob/dev/packages/core/src/module.js#L113

Maybe with a check like:
```
if (`handler.devModule` && `this.options._start`) {
  return
}
```

**Comment by pi0** [comment]
@atinux a non-dev module should never call `this.addModule(someDevModule)` 🙈

**Comment by galvez** [comment]
@Atinux good call, I'll update and add tests tonight.

**Comment by galvez** [comment]
> ```js
> if (handler.devModule && this.options._start) {
>   return
> }
> ```

@Atinux @pi0 For now I've added this bit:

```
    if (this.options.devModules.includes(handler) && this.options._start) {
      return
    }
```

Since we can't know if an external module is flagged with `devModule` without first requiring it. Advice welcome, but perhaps we should add the flag as part of the module improvements RFC?

-- Jonas

**Review by pi0 — DISMISSED** [review]
Good

**Comment by Atinux** [comment]
Failing tests:
screenshot 2019-02-28 at 11 30 07

You may want to add `devModules: []` before this line https://github.com/nuxt/nuxt.js/blob/dev/packages/config/src/config/_app.js#L25

😄

**Comment by galvez** [comment]
@Atinux weird, if you see the PR changelog, you'll see I added it already. I'll revisit now.

**Review by pi0 — APPROVED** [review]

**Pull request merged**